File: browser/js/core/parse/formulaContext.js
import { cmToUnit } from '../../utils.js';

// The named values a formula may read besides its own axis — twin of
// core/parse/formulaContext.hpp. Page fields are always cm and image fields always pixels;
// `unit` only picks the spelling PAGE_WIDTH / PAGE_HEIGHT report in.

// A blank formula is the identity: valid, and apply() returns its input unchanged.
export const isBlankFormula = (s) => !s || !s.trim();

// Validation has no live coordinates: an unbound axis probes at 1, as it always has.
export const withProbeAxes = (ctx) => ({
  ...ctx,
  x: Number.isFinite(ctx?.x) ? ctx.x : 1,
  y: Number.isFinite(ctx?.y) ? ctx.y : 1,
});

// `varName` first, so the context-free entry points keep their exact behaviour; then
// x, y, PAGE_WIDTH / PAGE_HEIGHT (+ _CM / _IN), IMAGE_WIDTH / IMAGE_HEIGHT. Case-sensitive.
// null for an unknown name and for a name whose field is absent or non-finite.
export const formulaConstant = (ctx, name, varName, varValue) => {
  if (name.length === 1 && name === varName) return varValue;
  const c = ctx || {};
  let v;
  if (name === 'x') v = c.x;
  else if (name === 'y') v = c.y;
  else if (name === 'PAGE_WIDTH') v = cmToUnit(c.pageWidthCm, c.unit);
  else if (name === 'PAGE_HEIGHT') v = cmToUnit(c.pageHeightCm, c.unit);
  else if (name === 'PAGE_WIDTH_CM') v = c.pageWidthCm;
  else if (name === 'PAGE_HEIGHT_CM') v = c.pageHeightCm;
  else if (name === 'PAGE_WIDTH_IN') v = cmToUnit(c.pageWidthCm, 'in');
  else if (name === 'PAGE_HEIGHT_IN') v = cmToUnit(c.pageHeightCm, 'in');
  else if (name === 'IMAGE_WIDTH') v = c.imageWidth;
  else if (name === 'IMAGE_HEIGHT') v = c.imageHeight;
  else return null;
  return Number.isFinite(v) ? v : null;
};

File: core/parse/formulaContext.cpp
#include "formulaContext.hpp"
#include "formulaParser.hpp"
#include <cmath>

namespace stencil::core {

  namespace {

    constexpr double CM_PER_INCH = 2.54;  // browser/js/utils/math.js CM_PER_INCH

    // Twin of cmToUnit(): only "in" converts, every other unit word reads as cm.
    double inUnit(double cm, const std::string& unit) {
      return unit == "in" ? cm / CM_PER_INCH : cm;
    }

    // Validation has no live coordinates: an unbound axis probes at 1, as it always has.
    FormulaContext withProbeAxes(FormulaContext c) {
      if (!std::isfinite(c.x)) c.x = 1.0;
      if (!std::isfinite(c.y)) c.y = 1.0;
      return c;
    }

  }  // namespace

  bool formulaConstant(const FormulaContext& ctx, const std::string& name, char varName,
                       double varValue, double& out) {
    if (name.size() == 1 && name[0] == varName) {
      out = varValue;
      return true;
    }
    double v = kFormulaUnset;
    if (name == "x") v = ctx.x;
    else if (name == "y") v = ctx.y;
    else if (name == "PAGE_WIDTH") v = inUnit(ctx.pageWidthCm, ctx.unit);
    else if (name == "PAGE_HEIGHT") v = inUnit(ctx.pageHeightCm, ctx.unit);
    else if (name == "PAGE_WIDTH_CM") v = ctx.pageWidthCm;
    else if (name == "PAGE_HEIGHT_CM") v = ctx.pageHeightCm;
    else if (name == "PAGE_WIDTH_IN") v = inUnit(ctx.pageWidthCm, "in");
    else if (name == "PAGE_HEIGHT_IN") v = inUnit(ctx.pageHeightCm, "in");
    else if (name == "IMAGE_WIDTH") v = ctx.imageWidth;
    else if (name == "IMAGE_HEIGHT") v = ctx.imageHeight;
    else return false;
    if (!std::isfinite(v)) return false;
    out = v;
    return true;
  }

  std::optional<double> FormulaParser::evaluate(const std::string& expr,
                                                const FormulaContext& ctx) {
    return evaluate(expr, '\0', 0.0, ctx);
  }

  bool FormulaParser::validate(const std::string& expr, const FormulaContext& ctx) {
    if (isBlankFormula(expr)) return true;
    return evaluate(expr, withProbeAxes(ctx)).has_value();
  }

  double FormulaParser::apply(const std::string& expr, char varName, double value,
                              bool allowFormulas, const FormulaContext& ctx) {
    if (!allowFormulas || isBlankFormula(expr)) return value;
    const auto result = evaluate(expr, varName, value, ctx);
    return result.has_value() ? *result : value;
  }

}
